File: common/src/main/java/dev/mayaqq/estrogen/client/cosmetics/CosmeticAnimation.java
package dev.mayaqq.estrogen.client.cosmetics;

import com.google.gson.JsonParser;
import com.mojang.datafixers.util.Pair;
import com.mojang.serialization.JsonOps;
import com.teamresourceful.resourcefullib.common.utils.files.GlobalStorage;
import dev.mayaqq.estrogen.Estrogen;
import dev.mayaqq.estrogen.client.cosmetics.model.animation.AnimationDefinition;
import io.netty.buffer.ByteBuf;

import java.io.File;
import java.io.Reader;
import java.nio.file.Path;
import java.util.Optional;
import java.util.concurrent.Callable;

public class CosmeticAnimation extends DownloadedAsset<AnimationDefinition> {

    public static final Path CACHE = GlobalStorage.getCacheDirectory(Estrogen.MOD_ID).resolve("cosmetics").resolve("animations");

    private volatile AnimationDefinition result;

    public CosmeticAnimation(String url) {
        super(CACHE, url);
    }

    public static CosmeticAnimation fromLocalFile(File file) {
        if(!file.isFile()) throw new IllegalArgumentException("File is not a file");
        CosmeticAnimation animation = new CosmeticAnimation("");
        animation.load(file, "");
        return animation;
    }

    public Optional<AnimationDefinition> getResult() {
        checkOrDownload();
        return Optional.ofNullable(result);
    }

    @Override
    protected void onLoad(AnimationDefinition definition) {
        this.result = definition;
    }

    @Override
    protected Optional<AnimationDefinition> read(Callable<Reader> readerSource) {
        try(Reader reader = readerSource.call()) {
            return AnimationDefinition.CODEC.parse(JsonOps.INSTANCE, JsonParser.parseReader(reader))
                .resultOrPartial(err -> EstrogenCosmetics.LOGGER.error("Failed to read cosmetic from url [{}]: {}", url, err));
        } catch (Exception e) {
            EstrogenCosmetics.LOGGER.error("Failed to load cosmetic from url [{}]", url, e);
            return Optional.empty();
        }
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/client/cosmetics/assets/CosmeticAnimation.kt
package dev.mayaqq.estrogen.client.cosmetics.assets

import com.google.gson.JsonParser
import com.mojang.serialization.Codec
import com.mojang.serialization.JsonOps
import dev.mayaqq.cynosure.client.models.animations.AnimationDefinition
import dev.mayaqq.estrogen.client.cosmetics.CACHE
import dev.mayaqq.estrogen.client.cosmetics.CosmeticAPI
import java.io.File
import java.io.Reader
import kotlin.jvm.optionals.getOrNull


class CosmeticAnimation(url: String) : DownloadAsset<AnimationDefinition, AnimationDefinition>(
    CACHE.resolve("animations"),
    url
) {

    override fun AnimationDefinition.onLoad() {
        result = this
    }

    override fun read(reader: () -> Reader): AnimationDefinition? {
        try {
            reader.invoke().use { reader ->
                return AnimationDefinition.CODEC.parse(JsonOps.INSTANCE, JsonParser.parseReader(reader))
                    .resultOrPartial{
                        CosmeticAPI.error("Failed to read cosmetic from url [{}]: {}", url, it)
                    }.getOrNull()
            }
        } catch (e: Exception) {
            CosmeticAPI.error("Failed to load cosmetic from url [{}]", url, e)
            return null
        }
    }

    companion object {
        val CODEC: Codec<CosmeticAnimation> = Codec.STRING.xmap(::CosmeticAnimation, CosmeticAnimation::url)

        fun fromLocalFile(file: File): CosmeticAnimation {
            require(file.isFile()) { "File is not a file" }
            val animation = CosmeticAnimation("")
            animation.load(file, "")
            return animation
        }
    }
}